lib/utils/flattenTokens.js

```diff
@@ -14,38 +14,50 @@
 import isPlainObject from 'is-plain-obj';
 
 /**
- * @typedef {import('../../types/DesignToken.d.ts').TransformedTokens} Tokens
- * @typedef {import('../../types/DesignToken.d.ts').TransformedToken} Token
+ * @typedef {import('../../types/DesignToken.ts').DesignToken} Token
+ * @typedef {import('../../types/DesignToken.ts').DesignTokens} Tokens
+ * @typedef {import('../../types/DesignToken.ts').TransformedToken} TransformedToken
+ * @typedef {import('../../types/DesignToken.ts').TransformedTokens} TransformedTokens
  */
 
 /**
  * @private
- * @param  {Tokens} slice - The plain object you want flattened into an array.
+ * @template {Token | TransformedToken} T
+ * @template {Tokens | TransformedTokens} U
+ * @param  {U} slice - The plain object you want flattened into an array.
  * @param {boolean} [usesDtcg] - Whether or not tokens are using DTCG syntax.
- * @param  {Token[]} [to_ret] - Tokens array. This function is recursive therefore this is what gets passed along.
- * @return {Token[]}
+ * @param  {Array<T>} [result] - Tokens array. This function is recursive therefore this is what gets passed along.
+ * @param  {string[]} [keyMemo] - Memo of keys we're traversing in order of depth
+ * @return {Array<T>}
  */
-function _flattenTokens(slice, usesDtcg, to_ret = []) {
-  for (let name in slice) {
-    if (Object.hasOwn(slice, name)) {
-      // TODO: this is a bit fragile and arbitrary to stop when we get to a 'value' property.
-      if (isPlainObject(slice[name]) && Object.hasOwn(slice[name], `${usesDtcg ? '$' : ''}value`)) {
-        to_ret.push(/** @type {Token} */ (slice[name]));
-      } else if (isPlainObject(slice[name])) {
-        _flattenTokens(slice[name], usesDtcg, to_ret);
+function _flattenTokens(slice, usesDtcg, result = [], keyMemo = []) {
+  for (let key in slice) {
+    if (Object.hasOwn(slice, key)) {
+      // Stop either when we encounter a "value" prop or if we find that every prop is not an object, meaning we cannot traverse any further
+      if (isPlainObject(slice[key]) && Object.hasOwn(slice[key], `${usesDtcg ? '$' : ''}value`)) {
+        result.push({
+          .../** @type {T} */ (slice[key]),
```

Comment:
@dbanksdesign  this change introduces top-level token copy which can broke some use cases. convertTokenData process tokens the same way

This can be broken change after 4.2.0 version

Reply:
`to_ret.push(/** @type {Token} */ (slice[name]));`

is equivalent to

`result.push({ .../** @type {T} */ (slice[key]) })`

given the token is an object type.. All we added is a key property to it.

I'm not sure why that would be a breaking change, could you elaborate or show an example?

Reply:
It's not the same, second is equivalent of  Object.assign({}, slice[key])

It's not breaking change for me, but for someone may be.

for example

```
for(const token of flattenTokens(dictionary.tokens)) {
  // ...some modification of token
}
```

in 4.2 will affect real tokens in dictionary
in 4.3 will ignore top properties change, for example token.value

Reply:
Hmm I see. The flattenTokens util should return a value that isn't coupled to the input value imo, so mutating the result should not mutate the input, this is unintentional and a bad practice if you ask me.

I would consider this an unintentional bugfix then; while people may rely on the "bug" (unintended behavior) in their code, it doesn't make the bugfix warrant a major bump.

Reply:
I understand that. I fixed it in my pipeline.

But to make function totally pure you should use full clone instead of just shallow clone.
Because current behavior allow of update for example subproperties of token, i.e attributes, properties of object values etc.

Reply:
True, I think in a lot of places I'm already using `structuredClone()` but we should verify for any function that is doing some kind of token handling that these functions are pure. https://github.com/amzn/style-dictionary/blob/main/lib/utils/flattenTokens.js#L62 is a good example where it's not yet being done.
Created an issue for this https://github.com/amzn/style-dictionary/issues/1412